web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts

```diff
@@ -317,11 +328,40 @@ export default class SubkeyPopup implements GestureHandler {
       // We're going to adjust the top of the box to ensure it stays
       // pixel aligned, otherwise we can get antialiasing artifacts
       // that look ugly
-      let top = Math.floor(keyRect.top - _BoxRect.top - 9 + delta);
+      let top = Math.floor(
+        Number.parseInt(cms.top, 10) +
+        Number.parseInt(cms.height, 10) +
+        // Padding is not included in content-box (or in content-box's top positioning)...
+        // but half the padding-top seems useful on all tested devices.
+        // Not sure exactly why.
+        Number.parseInt(cms.paddingTop, 10)/2 +
+        Number.parseInt(cms.paddingBottom, 10)
+      );
       ccs.top = top + 'px';
-      ccs.left = (keyRect.left - _BoxRect.left) + 'px';
-      ccs.width = keyRect.width + 'px';
-      ccs.height = (keyRect.bottom - _BoxRect.top - top - 1) + 'px'; //(height - 1) + 'px';
+
+      // Shorten the callout if the subkey menu is being constrained within WebView bounds, thus
+      // overlapping the base key.  Do so (mostly) proportionally to how much is obscured.
+      const targetHeight = (keyRect.height - delta) / 5 + STUB_HEIGHT;
+      const maxHeight = keyRect.bottom - _BoxRect.top - top - 1;
+      const selectedHeight = maxHeight < targetHeight ? maxHeight : targetHeight;
+      ccs.borderTopWidth = (selectedHeight) + 'px';
+
+      const STUB_RATIO = targetHeight > STUB_HEIGHT
+        ? targetHeight / (targetHeight - STUB_HEIGHT)
+        : 1;
+
+      const DESIRED_BASE_WIDTH = keyRect.width * STUB_RATIO;
+      const MAX_WIDTH = this.menuWidth - 2 * CORNER_ROUNDING;
+      const TARGET_BASE_WIDTH = MAX_WIDTH < DESIRED_BASE_WIDTH ? MAX_WIDTH : DESIRED_BASE_WIDTH;
+
+      const BASE_LEFT = (keyRect.left - _BoxRect.left - (TARGET_BASE_WIDTH - keyRect.width)/2);
+      const OVERRUN_RIGHT = Math.max(0, (BASE_LEFT + TARGET_BASE_WIDTH) - (_BoxRect.right - CORNER_ROUNDING));
+      const OVERRUN_LEFT =  Math.max(0, CORNER_ROUNDING-BASE_LEFT);
```

Comment:
I think these should all be `camelCased` because they are not eternal constants. Only fixed values should be `ALL_CAPS`, and those should generally be declared in a block at the top of the file.